polynote-kernel/src/main/scala/polynote/messages/Messages.scala

```diff
@@ -104,16 +127,92 @@ object RunCell extends MessageCompanion[RunCell](3)
 final case class CellResult(notebook: ShortString, id: TinyString, result: Result) extends Message
 object CellResult extends MessageCompanion[CellResult](4)
 
-final case class ContentEdit(pos: Int, length: Int, content: String) {
-  def applyTo(rope: Rope): Rope = if (length > 0) rope.delete(pos, length).insertAt(pos, Rope(content)) else rope.insertAt(pos, Rope(content))
+/**
+  * A ContentEdit at a position deletes some amount (possibly nothing) after that position, and then inserts some
+  * content (possibly empty) at that position.
+  *
+  * @param pos          The position at which the edit occurs
+  * @param deleteLength The number of characters to delete after the given position
+  * @param content      The content to insert
+  */
+final case class ContentEdit(pos: Int, deleteLength: Int, content: String) {
+  def applyTo(rope: Rope): Rope = if (deleteLength > 0) rope.delete(pos, deleteLength).insertAt(pos, Rope(content)) else rope.insertAt(pos, Rope(content))
+
+  // Given another edit which occured after this edit, create an edit that will be equivalent to applying this edit
+  // when applied after the given edit instead.
+  def rebase(other: ContentEdit): ContentEdit = other match {
+    // if the other edit is entirely before this edit, just shift this edit by the length delta
+    case ContentEdit(otherPos, otherDeleteLength, otherContent) if otherPos + otherDeleteLength <= pos =>
+      copy(pos = pos + otherContent.length - otherDeleteLength)
+
+    // if the other edit is entirely after this edit, nothing to do
+    case ContentEdit(otherPos, _, _) if otherPos >= pos + deleteLength => this
```

Comment:
Hmm, from what I can tell `ContentEditSpec` doesn't cover this case or the one above. Looks like it only hits the conflicting case (except for sometimes the property test will hit the other cases).

Reply:
Oops, you're right – will add cases to make sure all cases are covered.